test/jdk/com/sun/crypto/provider/Cipher/AEAD/GCMBufferTest.java

```diff
@@ -0,0 +1,547 @@
+                            plen, data, len + inofs);
+                        out = Arrays.copyOfRange(data, inofs,
+                            len + rlen + inofs);
+                    }
+                    case HEAP, DIRECT -> {
+                        rlen = cipher.doFinal(bbin, bbout);
+                        bbout.flip();
+                        out = new byte[bbout.remaining() - offset];
+                        bbout.position(offset);
+                        bbout.get(out);
+                    }
+                    default -> throw new Exception("Unknown op: " + v.name());
+                }
+                len += rlen;
+
+                if (len != output.length ||
+                    Arrays.compare(out, 0, len, output, 0,
+                        output.length) != 0) {
+                    throw new Exception("Ciphertext mismatch (" + v.name() +
+                        "):\nresult   (len=" + len + "):" +
+                        byteToHex(out) +
+                        "\nexpected (len=" + output.length + "):" +
+                        String.format("%0" + (output.length << 1) + "x",
+                            new BigInteger(1, output)));
+                }
+            }
+        }
+    }
+
+    String byteToHex(byte[] barray) {
+        StringBuilder s = new StringBuilder();
+        for (byte b : barray) {
+            s.append(String.format("%02x", b));
+        }
+        return s.toString();
+    }
+
+    public static void main (String args[]) throws Exception {
+
+        initTest();
+
+        new GCMBufferTest("AES/GCM/NoPadding", List.of(dtype.BYTE));
+        new GCMBufferTest("AES/GCM/NoPadding",
+            List.of(dtype.BYTE, dtype.BYTE));
+        new GCMBufferTest("AES/GCM/NoPadding", List.of(dtype.HEAP));
+        new GCMBufferTest("AES/GCM/NoPadding",
+            List.of(dtype.HEAP, dtype.HEAP));
+        new GCMBufferTest("AES/GCM/NoPadding", List.of(dtype.DIRECT));
+        new GCMBufferTest("AES/GCM/NoPadding",
+            List.of(dtype.BYTE, dtype.DIRECT), false);
+        new GCMBufferTest("AES/GCM/NoPadding",
+            List.of(dtype.DIRECT, dtype.BYTE), false);
+        new GCMBufferTest("AES/GCM/NoPadding",
+            List.of(dtype.BYTE, dtype.HEAP, dtype.DIRECT), false);
+    }
+
+    private static byte[] HexToBytes(String hexVal) {
+        if (hexVal == null) {
+            return new byte[0];
+        }
+        byte[] result = new byte[hexVal.length()/2];
+        for (int i = 0; i < result.length; i++) {
+            String byteVal = hexVal.substring(2*i, 2*i +2);
+            result[i] = Integer.valueOf(byteVal, 16).byteValue();
+        }
+        return result;
+    }
+
+    static void initTest() {
+        byte[] in = new byte[256];
+        byte[] inx = new byte[2075];
+        Arrays.fill(in, (byte) 0);
+        Arrays.fill(inx, (byte) 0);
+
+        datamap.put("AES/GCM/NoPadding", List.of(
+            new Data("AES",
```

Comment:
Just curious, is the data self-generated or got from some spec/external sources?

Reply:
The last one, in this diff, is generated by me.  The previous ones are KAT.